### PR TITLE
Add ECS bundle concept

### PR DESCRIPTION
**Pull request by Rhuagh**
Opening this early to get feedback. Is this a viable concept or should I trash it immediately?

### CODE REVIEW

**Comment by jojolepro** [comment]
@Rhuagh Basically they are code wrappers to make the code less heavy and reduce the boiler plate?
I'd like to know how that handles conflicts between dependencies and component registration too.

**Comment by Rhuagh** [comment]
Well, they're a generification of with_renderer in the ApplicationBuilder, so instead of adding functions there for every single concept (with_audio, with_transform), we can add bundles instead.

Conflicts are handled by specs for the most part. Resources are overridden, component registration is essentially a NOP for any duplicate registration. System registration is a panic if it already exists. When it comes to dependencies, each bundle handles that on it's own. Look at the TransformBundle for example, it has support for dependencies.

**Comment by Rhuagh** [comment]
An alternative would be to have the build() function take the ApplicationBuilder as argument and return it wrapped in a Result.

File: src/app.rs [hunk 1]
```diff
@@ -569,6 +570,38 @@ impl<'a, 'b, T: State + 'a> ApplicationBuilder<'a, 'b, T> {
...
+    ///
+    /// # Errors
+    ///
+    /// This function creates systems and resources, which use any number of dependent
+    /// crates or APIs, which could result in any number of errors.
+    /// See each individual bundle for a description of the errors it could produce.
+    ///
+    pub fn with_bundle<B>(mut self, bundle: B) -> Result<Self>
+    where
+        B: for<'c> ECSBundle<'a, 'b, (&'c EventsLoop)>,
```
Comment (omni-viral) [inline src/app.rs]: RenderSystem is special as it the only one that need access to `EventsLoop` during building.
What should we do if there will be new bundle that will require some other data from `Application` to build?

Comment (omni-viral) [inline src/app.rs]: Parentheses are unnecessary. If you want to make a tuple with single element you need to make it like this `(A,)`

Comment (Rhuagh) [reply src/app.rs]: Yeah, it's not perfect, but I haven't figured out a good solution to that yet.

Comment (Rhuagh) [reply src/app.rs]: Updated, the bundle trait now works on the ApplicationBuilder directly.

**Comment by Rhuagh** [comment]
Refactored to use ApplicationBuilder directly.

File: src/ecs/mod.rs [hunk 2]
```diff
@@ -7,14 +7,15 @@ pub mod transform;
...
-/// Extension trait for all systems.
-pub trait SystemExt<'a, A>: System<'a> {
-    /// Constructs a new system with the given arguments.
-    /// Registers all supported components with the World.
-    /// Puts resources into World.
-    fn build(args: A, world: &mut World) -> Result<Self>
-    where
-        Self: Sized;
+/// Describes a bundle of ECS components, resources and systems
+pub trait ECSBundle<'a, 'b, T: ::state::State + 'a> {
```
Comment (omni-viral) [inline src/ecs/mod.rs]: This bound shouldn't be here. Please, remove it from `ApplicationBuilder`'s functions where `T: State` is not required. I believe the only function that requires `T: State` is `ApplicationBuilder::build`

Comment (Rhuagh) [reply src/ecs/mod.rs]: Well, ApplicationBuilder requires it at the moment, and that's a separate change from this PR imho.

Comment (Rhuagh) [reply src/ecs/mod.rs]: I fixed it.

**Comment by Xaeroxe** [comment]
Looks good so far!  What would you consider the remaining tasks on this PR?

**Comment by Rhuagh** [comment]
A bit of cleanup, and making a bundle for the pong  stuff. And more documentation

**Comment by Rhuagh** [comment]
Ready to go after rereview.

**Comment by Xaeroxe** [comment]
Override.

File: src/app.rs [hunk 3]
```diff
@@ -585,65 +585,15 @@ impl<'a, 'b, T: State + 'a> ApplicationBuilder<'a, 'b, T> {
...
-        use ecs::SystemExt;
-        use ecs::rendering::RenderSystem;
-        let render_sys = RenderSystem::build((&self.events, pipeline, config), &mut self.world)?;
-        Ok(
-            self.with_thread_local(render_sys)
-                .register_mesh_asset()
-                .register_texture_asset()
-                .register_material_not_yet_asset(),
-        )
+    pub fn with_bundle<B>(self, bundle: B) -> Result<Self>
```
Comment (torkleyy) [inline src/app.rs]: If a bundle offers multiple systems, how would one manage dependencies? Should maybe the documentation list the system names it adds?

File: src/ecs/audio/bundle.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        &self,
+        mut builder: ApplicationBuilder<'a, 'b, T>,
+    ) -> Result<ApplicationBuilder<'a, 'b, T>> {
+        // Remove option here when specs get support for optional fetch in
+        // released version
+        if !builder.world.res.has_value(
+            ResourceId::new::<Option<Output>>(),
+        )
+        {
+            builder = builder.with_resource(default_output());
```
Comment (torkleyy) [inline src/ecs/audio/bundle.rs]: We should really have an entry API here.

File: src/ecs/audio/bundle.rs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        }
+
+        let dj = match *builder.world.read_resource::<Option<Output>>() {
```
Comment (torkleyy) [inline src/ecs/audio/bundle.rs]: Should use `map`.

File: src/ecs/audio/bundle.rs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            Some(ref audio_output) => Some(Dj::new(&audio_output)),
+
+            None => None,
+        };
+
+        if let Some(dj) = dj {
+            builder = builder.with_resource(dj).with(
```
Comment (torkleyy) [inline src/ecs/audio/bundle.rs]: That's where we need `add_*`.

File: src/ecs/input/bundle.rs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+use app::ApplicationBuilder;
+use config::Config;
+use ecs::ECSBundle;
+use ecs::input::{Bindings, InputHandler};
+use error::Result;
+
+/// Bundle for adding the InputHandler and input bindings
+///
+/// # Errors
```
Comment (torkleyy) [inline src/ecs/input/bundle.rs]: Should be `## Errors`

File: src/ecs/input/bundle.rs [hunk 8]
```diff
@@ -0,0 +1,48 @@
+//! ECS input bundle
+
+use app::ApplicationBuilder;
+use config::Config;
+use ecs::ECSBundle;
+use ecs::input::{Bindings, InputHandler};
+use error::Result;
+
+/// Bundle for adding the InputHandler and input bindings
```
Comment (torkleyy) [inline src/ecs/input/bundle.rs]: Should be `InputHandler`.

File: src/ecs/input/bundle.rs [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+impl InputBundle {
+    /// Create a new input bundle with no bindings
+    pub fn new() -> Self {
+        Self { bindings: None }
+    }
+
+    /// Use the provided bindings with the InputHandler
+    pub fn with_bindings(mut self, bindings: Bindings) -> Self {
```
Comment (torkleyy) [inline src/ecs/input/bundle.rs]: For the sake of a flatter syntax, could we have `new(Bindings)` and `from_file` instead? I wonder if we'd want no bindings at all.

File: src/ecs/input/bundle.rs [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        self.bindings = Some(bindings);
+        self
+    }
+
+    /// Load bindings from file
+    pub fn with_bindings_from_file(self, file: &str) -> Self {
```
Comment (torkleyy) [inline src/ecs/input/bundle.rs]: Should be using path API.

File: src/ecs/rendering/bundle.rs [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+use ecs::rendering::resources::{Factory, AmbientColor};
+use ecs::rendering::systems::RenderSystem;
+use ecs::transform::components::*;
+use error::{Error, Result};
+
+/// Rendering bundle
+///
+/// Will register all necessary components needed for rendering, along with any resources.
+/// Will also register asset contexts with the asset Loader, and add systems for merging
+/// AssetFutures into their respective component.
```
Comment (torkleyy) [inline src/ecs/rendering/bundle.rs]: Should be `AssetFuture`.

File: src/ecs/rendering/bundle.rs [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+///
+/// Will add RenderSystem as a thread local system.
```
Comment (torkleyy) [inline src/ecs/rendering/bundle.rs]: Should be `RenderSystem`

File: src/ecs/rendering/bundle.rs [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+///
+/// # Errors
```
Comment (torkleyy) [inline src/ecs/rendering/bundle.rs]: Should be `## Errors`

File: src/ecs/rendering/bundle.rs [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+///
+/// Returns errors related to:
+///
+/// * Renderer creation
```
Comment (torkleyy) [inline src/ecs/rendering/bundle.rs]: :+1:

File: src/ecs/rendering/bundle.rs [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+            .with_resource(Factory::new())
+            .with_resource(cam)
+            .with_resource(AmbientColor(Rgba::from([0.01; 3])))
+            .register::<Transform>()
+            .register::<LightComponent>()
+            .register::<MaterialComponent>()
+            .register::<MeshComponent>()
+            .register::<TextureComponent>();
+
+        // asset stuff, enable/disable flag for this?
```
Comment (torkleyy) [inline src/ecs/rendering/bundle.rs]: Please add a `FIXME` or `TODO` for these comments.

File: src/ecs/rendering/mod.rs [hunk 16]
```diff
@@ -1,9 +1,11 @@
 //! `amethyst` rendering ecs module
 
+pub use self::bundle::RenderBundle;
```
Comment (torkleyy) [inline src/ecs/rendering/mod.rs]: I think we should decide on whether we want to have a public module or reexported items. Given these bundles all have the same postfix, a conflict is impossible. I just see this inconsistency existed before the PR :worried:

File: src/ecs/transform/bundle.rs [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+use ecs::ECSBundle;
+use ecs::transform::*;
+use error::Result;
+
+/// Transform bundle
+///
+/// Will register transform components, and the TransformSystem.
+/// TransformSystem will be registered with name "transform_system".
+///
+/// # Errors
```
Comment (torkleyy) [inline src/ecs/transform/bundle.rs]: Same as above

File: src/ecs/transform/bundle.rs [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+///
+/// No errors will be returned by this bundle.
+///
+/// # Panics
```
Comment (torkleyy) [inline src/ecs/transform/bundle.rs]: Should be `## Panics`

File: src/ecs/transform/bundle.rs [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+///
+/// Panics in TransformSystem registration if the bundle is applied twice in the same dispatcher.
```
Comment (torkleyy) [inline src/ecs/transform/bundle.rs]: `TransformSystem`

File: src/ecs/transform/bundle.rs [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+impl<'a> TransformBundle<'a> {
+    /// Create a new transform bundle
+    pub fn new() -> Self {
+        Self { dep: &[] }
+    }
+
+    /// Set dependencies for the TransformSystem
+    pub fn with_dep(mut self, dep: &'a [&'a str]) -> Self {
```
Comment (torkleyy) [inline src/ecs/transform/bundle.rs]: This is something that makes the new bundle system more complex and gives the user less control. Also see my comment above. But well, let's see how it works out.

File: src/ecs/transform/bundle.rs (hunk 19 above)
Comment (Xaeroxe) [reply src/ecs/transform/bundle.rs]: That's what he wrote?

File: src/ecs/rendering/bundle.rs (hunk 12 above)
Comment (Xaeroxe) [reply src/ecs/rendering/bundle.rs]: That's what he wrote?

File: src/ecs/input/bundle.rs (hunk 8 above)
Comment (Xaeroxe) [reply src/ecs/input/bundle.rs]: That's what he wrote?

File: src/ecs/transform/bundle.rs (hunk 19 above)
Comment (torkleyy) [reply src/ecs/transform/bundle.rs]: Yep, just without backticks.

File: src/ecs/rendering/bundle.rs (hunk 12 above)
Comment (omni-viral) [reply src/ecs/rendering/bundle.rs]: Maybe he means absence of grave acent "`"?

File: src/app.rs (hunk 3 above)
Comment (Rhuagh) [reply src/app.rs]: I'd put all the information that the user could need in the documentation for each bundle.

File: src/ecs/audio/bundle.rs (hunk 4 above)
Comment (Rhuagh) [reply src/ecs/audio/bundle.rs]: Yes.

File: src/ecs/audio/bundle.rs (hunk 6 above)
Comment (Rhuagh) [reply src/ecs/audio/bundle.rs]: Yep.

File: src/ecs/audio/bundle.rs (hunk 5 above)
Comment (Rhuagh) [reply src/ecs/audio/bundle.rs]: Not sure I follow.

Comment (Rhuagh) [reply src/ecs/audio/bundle.rs]: Got it now.

File: src/ecs/input/bundle.rs (hunk 7 above)
Comment (Rhuagh) [reply src/ecs/input/bundle.rs]: I just followed what @csherratt has done with the ApplicationBuilder etc.

File: src/ecs/input/bundle.rs (hunk 9 above)
Comment (Rhuagh) [reply src/ecs/input/bundle.rs]: From what I've understood, the InputHandler only use bindings for actions/axes, but the user can also access the internal storages directly for a lot of things, so you aren't required to use bindings, which is why I made that a separate function.

File: src/ecs/transform/bundle.rs (hunk 20 above)
Comment (Rhuagh) [reply src/ecs/transform/bundle.rs]: You are not forced to use the bundle system, you can do all it does on it's own, and the code is open. 
I feel this is a way of lowering the threshold to start using the engine, without giving anything up really.

**Pull request merged**